nexus/db-queries/src/db/datastore/ip_pool.rs

```diff
@@ -324,6 +324,66 @@ impl DataStore {
             })
     }
 
+    pub async fn ip_pool_allocated_count(
+        &self,
+        opctx: &OpContext,
+        authz_pool: &authz::IpPool,
+    ) -> Result<u128, Error> {
+        opctx.authorize(authz::Action::Read, authz_pool).await?;
+
+        use db::schema::external_ip;
+
+        let count = external_ip::table
+            .filter(external_ip::ip_pool_id.eq(authz_pool.id()))
+            .filter(external_ip::time_deleted.is_null())
+            .select(external_ip::id)
+            .count()
+            .first_async::<i64>(&*self.pool_connection_authorized(opctx).await?)
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
+
+        // This is a bit strange but it should be ok.
+        //   a) SQL tends to represent counts as signed integers even though
+        //      they can't really be negative, and Diesel follows that. We
+        //      assume here that it will be a positive i64.
+        //   b) It could clearly be a u64, but the calling code immediate
+        //      converts it to a u128 anyway, so it would be pointless.
+        u128::try_from(count).map_err(|_e| {
+            Error::internal_error(&format!(
+                "Failed to convert i64 {} SQL count to u64",
+                count
+            ))
+        })
+    }
+
+    // TODO: should this just return the vec of ranges?
+    // TODO: generally we never retrieve all of anything, how bad is that
+    pub async fn ip_pool_total_capacity(
+        &self,
+        opctx: &OpContext,
+        authz_pool: &authz::IpPool,
+    ) -> Result<u128, Error> {
+        opctx.authorize(authz::Action::Read, authz_pool).await?;
+        opctx.authorize(authz::Action::ListChildren, authz_pool).await?;
+
+        use db::schema::ip_pool_range;
+
+        let ranges: Vec<IpPoolRange> = ip_pool_range::table
```

Comment:
Interesting question! It hadn't occurred to me to give separate `allocated` and `capacity` counts for v4 and v6, but it makes a lot of sense, doesn't it. I will think about doing that here — it's always nice to avoid followup API changes where we can. One nice thing about that is that the IPv4 counts can be `u32`s.

Another way we could do this would be to have a given pool be IPv4 or IPv6 only, but I don't really like that. Right now you can add both IPv4 and IPv6 ranges to a pool, and it would be kind of complicated and/or confusing to require pools to have only one or the other.

Reply:
All right, check out 39edb8ffad89148e40f7be298096e6b7aa5ac899. A little complicated, maybe, but the end result is 

```
{
  ipv4: { allocated, capacity },
  ipv6: { allocated, capacity },
}
```

Sound good?

Reply:
I think so, thanks for doing this! This also helps a lot since external v4s can be quite constrained in practice.